include/swift/IDE/CodeCompletionResultType.h

```diff
@@ -107,54 +109,191 @@ enum class CodeCompletionResultTypeRelation : uint8_t {
+class USRBasedTypeContext {
+public:
+  class ContextualType {
+    /// The types that the result type must be identical/convertible to to
+    /// compute the type relation. Needs to be a vector because for the
+    /// conextual type `some MyProto & MyOtherProto`, the return type must be
+    /// convertible to both \c MyProto and \c MyOtherProto to be considered
+    /// convertible.
+    llvm::SmallVector<const USRBasedType *, 1> Types;
+
+    /// Whether a match against this type should be considered convertible
+    /// instead of identical. This is used to model optional conversions.
+    /// If we have
+    ///   let x: Int? = #^COMPLETE^#
+    /// we add both 'Int?' and 'Int' as contextual types to the
+    /// USRBasedTypeContext, but an identical match against 'Int' should only
+    /// be considered convertible.
+    bool IsConvertible;
+
+  public:
+    /// Compute the type relation of \p ResultType to this conextual type.
+    CodeCompletionResultTypeRelation
+    typeRelation(const USRBasedType *ResultType,
+                 const USRBasedType *VoidType) const;
+
+    ContextualType(ArrayRef<const USRBasedType *> Types, bool IsConvertible)
+        : Types(Types.begin(), Types.end()), IsConvertible(IsConvertible) {
+      assert(!Types.empty() && "USRBasedTypeContext::ContextualType should "
+                               "contain at least one type");
+    }
+  };
+
+private:
+  /// Cache of the \c Void type because its regularly used in type relation
+  /// computation.
+  const USRBasedType *VoidType;
+
+  SmallVector<ContextualType, 4> ContextualTypes;
+
+public:
+  /// Create a USR-based equivalent of the \p TypeContext.
+  USRBasedTypeContext(const ExpectedTypeContext *TypeContext,
+                      const DeclContext *DC, USRBasedTypeArena &Arena);
+
+  /// Type relation of \p ResultType to this context.
+  CodeCompletionResultTypeRelation
+  typeRelation(const USRBasedType *ResultType) const;
+};
+
+/// A type that is purely USR-based and thus independent of an \c ASTContext.
+/// Its identity is defined by its USR. The object also stores the USRs of all
+/// its known supertypes (its superclasses and protocol conformances declared
+/// within the same module) so it is able to determine whether one type is
+/// convertible to another.
+/// Because the type is purely based on USRs, it can be serialized into the code
+/// completion cache and read from it later.
+/// \c USRBasedTypes are always allocated within a \c USRBasedTypeArena. Types
+/// in the same arena that have the same USR are represented by the same object.
+/// \c USRBasedTypes always represent canonical types so that equivalent types
+/// are represented by the same USR.
+/// A type with an empty USR represents the null type.
+class USRBasedType {
+  std::string USR;
+
+  /// The superclasses of the type and all protocol conformances that are
+  /// declared in the same module that the type was defined.
+  std::vector<const USRBasedType *> Supertypes;
```

Comment:
This type is allocated on a BumpAllocator, right? If so we can't use non-trivial type fields (std::string and std::vector). Since the allocator doesn't know its contents, it doesn't destruct the objects even if the allocator itself is destructed.

These should be `StringRef` and `ArrayRef<USRBaseType *>` those storage is the same allocator as USRBaseType.

Reply:
Good catch. I didn’t think about that.